snapcraft/internal/elf.py

```diff
@@ -292,25 +295,29 @@ def _extract(self, path: str) -> ElfDataTuple:  # noqa: C901
                 elf.header.e_machine,
             )
 
+            for segment in elf.iter_segments():
+                if segment["p_type"] == "PT_DYNAMIC":
+                    is_dynamic = True
+                    for tag in segment.iter_tags("DT_NEEDED"):
+                        needed = _ensure_str(tag.needed)
+                        libs[needed] = NeededLibrary(name=needed)
+                    for tag in segment.iter_tags("DT_SONAME"):
+                        soname = _ensure_str(tag.soname)
+                elif segment["p_type"] == "PT_GNU_STACK":
+                    # p_flags holds the bit mask for this segment.
+                    # See `man 5 elf`.
+                    mode = segment["p_flags"]
+                    if mode & elftools.elf.constants.P_FLAGS.PF_X:
+                        execstack_set = True
+                elif segment["p_type"] == "PT_INTERP":
```

Comment:
It's effectively equivalent after looking at the pyelf code, but I think it should key off the instance types rather than checking the p_type.

e.g. `if isinstance(segment, InterpSegment)` or `if isinstance(segment, NoteSegment)`.

I'm sure the library is sane, but the code here for these two segments are using the interfaces provided by those types.

Reply:
It looks like there are special `InterpSegment`, `DynamicSegment`, and `NoteSegment` classes, but the `PT_GNU_STACK` segment shows up as a generic `elftools.elf.segments.Segment` instance.

I can make the change, but it would only cover three of the four branches of the if statement.  That's not obviously a readability win, although I guess it could help with static analysis?

Reply:
I will say the code is effectively OK either way (based on https://github.com/eliben/pyelftools/blob/master/elftools/elf/elffile.py#L452). 

I think it would help analysis. 

But I think if you are going to use the interfaces of specific subclass type in python, I would ensure it's an instance of it (say the library incorrectly returned a Segment instead of NoteSegment for whatever reason...).  And while the readability might look a little more awkward at a glance, it's more obvious that you are now interfacing with a more-specific subclass, and provides the developer with a strong hint as to where to find iter_notes(), get_interp_name(), iter_tags(), etc. which are not in the base Segment class.  I'm not familiar with pyelf, so I had to look it up to and sort out why these calls were OK as-is.

FWIW pyelf's example scripts rely on this approach: https://github.com/eliben/pyelftools/blob/master/examples/elf_notes.py#L25

It's redundant, but an option to keep it both ways is:
`if segment["p_type"] == "PT_DYNAMIC" and isinstance(segment, DynamicSegment):`